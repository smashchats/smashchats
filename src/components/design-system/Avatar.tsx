import React, { type PropsWithChildren } from "react";
import { View } from "react-native";

type Props = PropsWithChildren<{
    flex?: number;
    bg?: string;
    bgColor?: string;
    borderRadius?: number;
    size?: number;
}>;

export function Avatar({
    children,
    flex,
    bg,
    bgColor,
    borderRadius,
    size,
}: Props): JSX.Element {
    return (
        <View
            style={{
                backgroundColor: bg ?? bgColor,
                flex,
                minWidth: size ?? 64,
                maxWidth: size ?? 64,
                width: size ?? 64,
                minHeight: size ?? 64,
                maxHeight: size ?? 64,
                height: size ?? 64,
                borderRadius: borderRadius ?? 16,
                overflow: "hidden",
            }}
        >
            {children}
        </View>
    );
}
