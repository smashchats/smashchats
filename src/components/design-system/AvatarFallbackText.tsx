import React from "react";
import { Text, View } from "react-native";

type Props = {
    name: string;
};

export function AvatarFallbackText({ name }: Readonly<Props>): JSX.Element {
    return (
        <View
            style={{
                alignItems: "center",
                justifyContent: "center",
                height: "100%",
                width: "100%",
            }}
        >
            <Text style={{ fontSize: 24, color: "white", fontWeight: "bold" }}>
                {(name ?? "")
                    .split(" ")
                    .map((word) => word[0])
                    .join("")
                    .toUpperCase()}
            </Text>
        </View>
    );
}
