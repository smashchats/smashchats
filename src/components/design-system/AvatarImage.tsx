import React, { type PropsWithChildren } from "react";
import { ViewStyle } from "react-native";
import { Image, ImageSource } from "expo-image";

type Props = PropsWithChildren<
    {
        source?:
            | ImageSource
            | string
            | number
            | ImageSource[]
            | string[]
            | null;
        alt?: string;
        size?: number;
    } & ViewStyle
>;

export function AvatarImage({ source, alt, size }: Props): JSX.Element {
    return (
        <Image
            style={{
                minWidth: size ?? 64,
                maxWidth: size ?? 64,
                width: size ?? 64,
                minHeight: size ?? 64,
                maxHeight: size ?? 64,
                height: size ?? 64,
            }}
            source={source}
            alt={alt}
            contentFit={"contain"}
            transition={1000}
        />
    );
}
