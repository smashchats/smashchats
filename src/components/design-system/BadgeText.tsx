import React, { type PropsWithChildren } from "react";
import { DimensionValue, Text, TextStyle, View } from "react-native";

type Props = PropsWithChildren<
    {
        bg?: string;
        h?: DimensionValue;
    } & TextStyle
>;

export function BadgeText({
    children,
    flex,
    bg,
    h,
    color,
    ...rest
}: Props): JSX.Element {
    return (
        <View style={{ backgroundColor: bg, height: h, flex }}>
            <Text
                style={{
                    color: color ?? "white",
                    fontFamily:
                        'Cairo, apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif',
                    fontSize: 12,
                    textTransform: "uppercase",
                    ...rest,
                }}
            >
                {children}
            </Text>
        </View>
    );
}
