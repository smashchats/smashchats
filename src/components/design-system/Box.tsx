import React, { type PropsWithChildren } from "react";
import { DimensionValue, View, ViewStyle } from "react-native";

type Props = PropsWithChildren<
    {
        bg?: string;
        h?: DimensionValue;
    } & ViewStyle
>;

export function Box({ children, bg, h, ...rest }: Props): JSX.Element {
    return (
        <View
            style={{
                backgroundColor: bg,
                height: h,
                ...rest,
            }}
        >
            {children}
        </View>
    );
}
