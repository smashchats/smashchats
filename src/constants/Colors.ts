/**
 * Below are the colors that are used in the app. The colors are defined in the light and dark mode.
 * There are many other ways to style your app. For example, [Nativewind](https://www.nativewind.dev/), [Tamagui](https://tamagui.dev/), [unistyles](https://reactnativeunistyles.vercel.app), etc.
 */

const tintColorLight = '#0a7ea4';
const tintColorDark = '#fff';

export const Colors = {
    light: {
        text: '#11181C',
        tint: tintColorLight,
        icon: '#687076',
        tabIconDefault: '#687076',
        tabIconSelected: tintColorLight,

        background: '#222222',
        purple: '#64189b',
        darkGray: '#3F3F3F',
        textGray: '#696969',
        textLightGray: '#9B9B9B',
    },
    dark: {
        text: '#ECEDEE',
        tint: tintColorDark,
        icon: '#9BA1A6',
        tabIconDefault: '#9BA1A6',
        tabIconSelected: tintColorDark,

        background: '#222222',
        purple: '#64189b',
        darkGray: '#3F3F3F',
        textGray: '#696969',
        textLightGray: '#9B9B9B',
    },
    background: '#222222',
    purple: '#64189b',
    darkGray: '#3F3F3F',
    textGray: '#696969',
    textLightGray: '#9B9B9B',
    textWhite: '#F0F0F0',
};
