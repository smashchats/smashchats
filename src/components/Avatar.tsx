import { Colors } from "@/src/constants/Colors";
import { Avatar as AvatarComponent } from "@/src/components/design-system/Avatar";
import { AvatarImage } from "@/src/components/design-system/AvatarImage";
import { AvatarFallbackText } from "@/src/components/design-system/AvatarFallbackText";
import { Contact } from "@/src/models/Contacts";

type AvatarProps = {
    contact: Contact & { trusted_name: string | undefined };
    variant?: "small" | "large" | "xlarge";
};

export const Avatar = ({ contact, variant = "small" }: AvatarProps) => {
    const dimensions = {
        small: 40,
        large: 64,
        xlarge: 128,
    };
    const radiusDimensions = {
        small: 10,
        large: 16,
        xlarge: 32,
    };

    if (!contact) {
        return null;
    }

    const hasImage =
        contact.meta_avatar?.startsWith("data:image/png;base64,") ||
        contact.meta_avatar?.startsWith("data:image/jpeg;base64,") ||
        contact.meta_avatar?.startsWith("https://kinkverse.org") ||
        contact.meta_avatar?.startsWith("https://i.ytimg.com") ||
        contact.meta_avatar?.startsWith("https://upload.wiki") ||
        contact.meta_avatar?.startsWith("file://") ||
        contact.meta_avatar?.startsWith("https://github.com") ||
        contact.meta_avatar?.startsWith(
            "https://unstaticlabs.com/unstatic-logo.svg"
        );

    const avatarSize = dimensions[variant];

    return (
        <AvatarComponent
            bgColor={Colors.purple}
            borderRadius={radiusDimensions[variant]}
            size={avatarSize}
        >
            {!hasImage && (
                <AvatarFallbackText
                    name={
                        contact.trusted_name ??
                        contact.meta_title ??
                        "unnamed contact"
                    }
                />
            )}
            {hasImage && (
                <AvatarImage
                    alt={`${
                        contact.trusted_name ??
                        contact.meta_title ??
                        "unnamed contact"
                    }'s avatar'`}
                    borderRadius={radiusDimensions[variant]}
                    size={avatarSize}
                    source={contact.meta_avatar}
                />
            )}
        </AvatarComponent>
    );
};
