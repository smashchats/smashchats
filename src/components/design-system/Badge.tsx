import { Colors } from "@/src/constants/Colors.js";

import React, { type PropsWithChildren } from "react";
import { DimensionValue, View, StyleSheet, ViewStyle } from "react-native";

type Props = PropsWithChildren<
    {
        flex?: number;
        type?: string;
        bg?: string;
        bgColor?: string;
        size?: string;
        h?: DimensionValue;
    } & ViewStyle
>;

export function Badge({
    children,
    size,
    flex,
    bg,
    h,
    width,
    marginHorizontal,
    marginRight,
    paddingHorizontal,
    type,
    borderRadius,
    borderColor,
    bgColor,
    borderWidth,
    ...rest
}: Props): JSX.Element {
    return (
        <View
            style={{
                flexDirection: "row",
                alignItems: "center",
                justifyContent: "center",
                borderRadius: 16,
                paddingHorizontal: 8,
                paddingVertical: 1,

                height: h,
                flex,
                width,
                marginHorizontal,
                marginRight,

                ...styles.unselected,
                ...(type === "unselected" ? styles.unselected : {}),
                ...(type === "trusted" ? styles.trusted : {}),
                ...(type === "smashed" ? styles.smashed : {}),

                ...(borderWidth && { borderWidth }),
                ...(borderColor && { borderColor }),
                ...((bg || bgColor) && { backgroundColor: bg ?? bgColor }),
                ...(borderRadius && { borderRadius }),
                ...rest,
            }}
        >
            {children}
        </View>
    );
}

const styles = StyleSheet.create({
    smashed: {
        backgroundColor: Colors.purple,
        borderColor: Colors.purple,
        borderWidth: 2,
        // marginLeft: 30
    },
    trusted: {
        backgroundColor: "#3f3f3f",
        borderColor: "#3f3f3f",
        borderWidth: 2,
    },
    unselected: {
        backgroundColor: Colors.background,
        borderColor: Colors.purple,
        borderWidth: 2,
    },
});
