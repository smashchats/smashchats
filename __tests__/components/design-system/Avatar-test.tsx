import { render } from "@testing-library/react-native";
import { Avatar } from "@/src/components/design-system/Avatar.jsx";
import { AvatarImage } from "@/src/components/design-system/AvatarImage";
import { Colors } from "@/src/constants/Colors";
import { AvatarFallbackText } from "@/src/components/design-system/AvatarFallbackText";

describe("Avatar", () => {
    test("Renders correctly", () => {
        const tree = render(<Avatar size={100} />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    test("Renders correctly with image", () => {
        const tree = render(
            <Avatar bgColor={Colors.purple}>
                <AvatarImage
                    alt={`contact name's profile picture`}
                    borderRadius={16}
                    size={64}
                    source={"data:base64,image"}
                />
            </Avatar>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    test("Renders correctly with fallback text", () => {
        const tree = render(
            <Avatar bgColor={Colors.purple}>
                <AvatarFallbackText name={"contact name"} />
            </Avatar>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
