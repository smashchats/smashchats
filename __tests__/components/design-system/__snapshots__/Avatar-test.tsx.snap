// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": undefined,
      "borderRadius": 16,
      "flex": undefined,
      "height": 100,
      "maxHeight": 100,
      "maxWidth": 100,
      "minHeight": 100,
      "minWidth": 100,
      "overflow": "hidden",
      "width": 100,
    }
  }
/>
`;

exports[`Avatar Renders correctly with fallback text 1`] = `
<View
  style={
    {
      "backgroundColor": "#64189b",
      "borderRadius": 16,
      "flex": undefined,
      "height": 64,
      "maxHeight": 64,
      "maxWidth": 64,
      "minHeight": 64,
      "minWidth": 64,
      "overflow": "hidden",
      "width": 64,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "height": "100%",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 24,
          "fontWeight": "bold",
        }
      }
    >
      CN
    </Text>
  </View>
</View>
`;

exports[`Avatar Renders correctly with image 1`] = `
<View
  style={
    {
      "backgroundColor": "#64189b",
      "borderRadius": 16,
      "flex": undefined,
      "height": 64,
      "maxHeight": 64,
      "maxWidth": 64,
      "minHeight": 64,
      "minWidth": 64,
      "overflow": "hidden",
      "width": 64,
    }
  }
>
  <ViewManagerAdapter_ExpoImage
    accessibilityLabel="contact name's avatar"
    containerViewRef={"[React.ref]"}
    contentFit="contain"
    contentPosition={
      {
        "left": "50%",
        "top": "50%",
      }
    }
    height={64}
    maxHeight={64}
    maxWidth={64}
    minHeight={64}
    minWidth={64}
    nativeViewRef={"[React.ref]"}
    onError={[Function]}
    onLoad={[Function]}
    onLoadStart={[Function]}
    onProgress={[Function]}
    placeholder={[]}
    source={
      [
        {
          "uri": "data:base64,image",
        },
      ]
    }
    style={
      {
        "height": 64,
        "maxHeight": 64,
        "maxWidth": 64,
        "minHeight": 64,
        "minWidth": 64,
        "width": 64,
      }
    }
    transition={
      {
        "duration": 1000,
      }
    }
    width={64}
  />
</View>
`;
